pprust: adjust mixed comment printing and add regression test for #74745

davidtwco opened the pull request:
Fixes #74745.

This PR adds a regression test for #74745. While a `ignore-tidy-trailing-lines` header is required, this doesn't stop the test from reproducing, so long as there is no newline at the end of the file.

However, adding the header comments made the test fail due to a bug in pprust - so this PR also  adjusts the pretty printing of mixed comments so that the initial zero-break isn't emitted at the beginning of the line. Through this, the `block-comment-wchar` test can have the `pp-exact` file removed, as it no longer converges from pretty printing of the source.

davidtwco committed:
tests: add regression test for #74745

This commit adds a regression test for #74745. While a
`ignore-tidy-trailing-lines` header is required, this doesn't stop the
test from reproducing, so long as there is no newline at the end of the
file.

However, adding the header comments made the test fail due to a bug in
pprust, fixed in the previous commit.

Signed-off-by: David Wood <david@davidtw.co>

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

petrochenkov commented on src/test/pretty/block-comment-wchar.rs:
```diff
@@ -3,7 +3,6 @@
 // Test via: rustc --pretty normal src/test/pretty/block-comment-wchar.rs
 // ignore-tidy-cr
 // ignore-tidy-tab
-// pp-exact:block-comment-wchar.pp
```
Could you update `block-comment-wchar.pp` instead of removing it?

davidtwco replied:
Fixed!

davidtwco committed:
pprust: adjust mixed comment printing

This commit adjusts the pretty printing of mixed comments so that the
initial zero-break isn't emitted at the beginning of the line. Through
this, the `block-comment-wchar` test can have the `pp-exact` file
removed, as it no longer converges from pretty printing of the source.

Signed-off-by: David Wood <david@davidtw.co>

davidtwco committed:
compiletest: print diff for pretty tests

This commit modifies compiletest so that a diff of actual and expected
output is shown for pretty tests. This makes it far easier to work out
what has changed.

Signed-off-by: David Wood <david@davidtw.co>

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 1530563bd7a3880eec14d78e4a715c1bc4e24074 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 1530563bd7a3880eec14d78e4a715c1bc4e24074 -->
<!-- homu: {"type":"Approved","sha":"1530563bd7a3880eec14d78e4a715c1bc4e24074","approver":"petrochenkov"} -->
